File: leetcode/500. Keyboard Row/main.go
package main

import "strings"

func findWords(words []string) []string {
	letters := make(map[rune]int)

	keyboard := []struct {
		row string
		idx int
	}{
		{"qwertyuiop", 1},
		{"asdfghjkl", 2},
		{"zxcvbnm", 3},
	}

	for _, entry := range keyboard {
		for _, ch := range entry.row {
			letters[ch] = entry.idx
		}
	}

	result := make([]string, 0, len(words))

	for _, word := range words {
		if len(word) == 0 {
			continue
		}

		lower := strings.ToLower(word)
		row, canBeTyped := letters[rune(lower[0])], true

		for _, v := range lower {
			if letters[v] != row {
				canBeTyped = false
				break
			}
		}

		if canBeTyped {
			result = append(result, word)
		}
	}

	return result
}

File: leetcode/500. Keyboard Row/main.rs
impl Solution {
    pub fn find_words(words: Vec<String>) -> Vec<String> {
        let mut letters: HashMap<char, i32> = HashMap::new();

        for c in "qwertyuiop".to_string().chars() {
            letters.insert(c, 1);
        }

        for c in "asdfghjkl".to_string().chars() {
            letters.insert(c, 2);
        }

        for c in "zxcvbnm".to_string().chars() {
            letters.insert(c, 3);
        }

        words
            .into_iter()
            .filter(|word| {
                let lowercased_word = word.to_lowercase();
                let mut chars = lowercased_word.chars();

                if let Some(first_char) = chars.next() {
                    let row = letters.get(&first_char).unwrap_or(&0);
                    chars.all(|c| letters.get(&c).unwrap_or(&0) == row)
                } else {
                    false
                }
            })
            .collect()
    }
}

